src/net/connect.cpp

```diff
@@ -28,6 +30,30 @@ void mk_bufferevent_on_event(bufferevent *bev, short what, void *ptr) {
     delete cb;
 }
 
+void mk_bufferevent_on_event_ssl(bufferevent *bev, short what, void *ptr) {
+    // XXX: Maybe this can be refactored a bit to have less duplication with
+    // the above function.
```

Comment:
To do that, which I think would be neat, I suggest to refactor `mk_bufferevent_on_event` such that, in the error paths, the bufferevent is freed after the callbacks are called and to pass such bufferevents to the callee for further inspection. With this change in, you can refactor the piece of code that checks for invalid certificate to stay in the callback (i.e., rather than just checking `if (err)` you can also check if more specifically there was an openssl error and, in such case, be more informative with the caller.